File: 2021/2/solve.py
INPUT_FILE = '2021/2/input.txt'

from functools import reduce
from operator import add

def input(name = INPUT_FILE):
    with open(name) as f:
        for l in f:
            yield l.strip('\n')

CMDS1 = {
    'forward': lambda x: (x, 0),
    'up': lambda x: (0, -x),
    'down': lambda x: (0, x),
}

CMDS2 = {
    'forward': lambda a, x: (x, a*x, 0),
    'up': lambda _, x: (0, 0, -x),
    'down': lambda _, x: (0, 0, x),
}

def addv(a, b):
    return tuple(map(add, a , b))

def part1(i):
    x, y = reduce(
        addv,
        map(lambda c: CMDS1[c[0]](int(c[1])), i),
        (0, 0)
    )
    print(x*y)

def step(s, c):
    c, t = c
    _, _, a = s
    return addv(s, c(a, t))

def part2(i):
    x, y, _ = reduce(
        step,
        map(lambda c: (CMDS2[c[0]], int(c[1])), i),
        (0, 0, 0)
    )
    print(x*y)

def solve():
    i = list(map(str.split, input()))
    part1(i)
    part2(i)

solve()

File: 2021/2/solve.cpp
// requires -std=c++17

#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <tuple>
#include <map>
using namespace std;

enum DIRECTION
{
    FORWARD,
    UP,
    DOWN
};

vector<tuple<int, int>> readinput()
{
    vector<tuple<int, int>> input;
    string filename = "input.txt";
    string dirstr;
    int d, x;
    map<string, int> m {
        {"forward", FORWARD},
        {"up", UP},
        {"down", DOWN},
    };

    {
        ifstream file(filename, ios::in);
        while(file >> dirstr >> x)
        {
            int d = m[dirstr];
            input.push_back(tuple<int, int>(d, x));
        }
    }

    return input;
}

void part1(vector<tuple<int, int>>& input)
{
    int u=0, v=0;
    for(auto [d, x]: input)
    {
        switch(d)
        {
            case FORWARD: u+=x; break;
            case UP: v-=x; break;
            case DOWN: v+=x; break;
        }
    }

    cout << u*v << endl;
}

void part2(vector<tuple<int, int>>& input)
{
    int u=0, v=0, aim=0;
    for(auto [d, x]: input)
    {
        switch(d)
        {
            case FORWARD: u+=x; v+=aim*x; break;
            case UP: aim-=x; break;
            case DOWN: aim+=x; break;
        }
    }

    cout << u*v << endl;
}

int main()
{
    vector<tuple<int, int>> input = readinput();
    part1(input);
    part2(input);

    return 0;
}
